Contribute: update style guide to google + (minimal) exceptions

tstastny opened the pull request:
Updates style guide documentation with link to google, and out lines our exceptions.

@MaEtUgR @dagar please let me know if I got something wrong.

We would probably also want to add something about clang format / tidy once that's all up and running. But for now this at least gives the general direction what we're trying to move to.

Let me know what you think ;) 

NOTE: clang format will be used where astyle was before, we will need to remember to update this doc with instructions for e.g. the hook setup and auto formatting make command (does it stay the same? `make format` ?) @MaEtUgR @hamishwillee 
^see PX4/PX4-Autopilot#19855

MaEtUgR commented on en/contribute/code.md:
```diff
@@ -16,26 +16,66 @@ Changes will be merged when they pass our [continuous integration](https://en.wi
+public:
+
+    /**
+     * @breif Description of what this function does.
```
```suggestion
     * @brief Description of what this function does.
```

tstastny commented on en/contribute/code.md:
```diff
@@ -16,26 +16,66 @@ Changes will be merged when they pass our [continuous integration](https://en.wi
+
+- Source files use extension `*.cpp` instead of `*.cc`
+
+- An exception to the rule above are the MAVLink streams in [src/modules/mavlink/streams](https://github.com/PX4/PX4-Autopilot/tree/main/src/modules/mavlink/streams) which are ALL_UPPERCASE.hpp matching the MAVLink message name.
```
ah oops i copy pasted that bullet without modifying for the new context - this was originally referring to the old guide: https://github.com/PX4/PX4-user_guide/blob/main/en/contribute/code.md#file-name-conventions -- i thought the issue may be that these are autogenerated with that ALL_UPPERCASE or somehow tied to the mavlink repo (so requiring changes there as well), but if we can change this without issue, then definitely in for syncing with the google style and removing this bullet

hamishwillee commented on en/contribute/code.md:
````diff
@@ -16,26 +16,66 @@ Changes will be merged when they pass our [continuous integration](https://en.wi
+
+### Example Code Snippet
+
+```
````
If you use  ```cpp you get nicer code formatting.

hamishwillee commented on en/contribute/code.md:
```diff
@@ -57,6 +97,7 @@ Currently we have two types of source-based documentation:
   :::
   
   - Do not add documentation that can trivially be assumed from C++ entity names.
+  - ALWAYS specify units of variables, constants, and input/return parameters where they are defined
```
```suggestion
  - ALWAYS specify units of variables, constants, and input/return parameters where they are defined.
```

hamishwillee commented:
Nice. So we're no longer defining our coding standard by a tool (astyle)? If there is a toolchain that manages this it would be good to at least mention it. And if the intent is that clang will do it one day then putting that in a note with a link to the issue PR in which that is addressed would be even better - because it allows easier later cleanup. 

Do we need to say something specific about a dislike for magic numbers?

@dagar LGTM - if you're happy just merge.

tstastny commented:
> Do we need to say something specific about a dislike for magic numbers?

Added :) 

> So we're no longer defining our coding standard by a tool (astyle)? If there is a toolchain that manages this it would be good to at least mention it. And if the intent is that clang will do it one day then putting that in a note with a link to the issue PR in which that is addressed would be even better - because it allows easier later cleanup.

Yes clang format is intended to take this over (and eventually clang tidy doing even more heavy lifting) - I added a note/link in the PR description. @MaEtUgR if you think we could already add something about this to this document - please feel free to commit directly to this PR.

> NOTE: clang format will be used where astyle was before, we will need to remember to update this doc with instructions for e.g. the hook setup and auto formatting make command (does it stay the same? make format ?) @MaEtUgR @hamishwillee
^see https://github.com/PX4/PX4-Autopilot/pull/19855

hamishwillee commented:
Thanks @tstastny 

This looks good to me. I'm going to merge and we can iterate.
